pkg/ctl/create/nodegroup.go

```diff
@@ -0,0 +1,128 @@
+package create
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/kubicorn/kubicorn/pkg/logger"
+	"github.com/spf13/cobra"
+	"github.com/weaveworks/eksctl/pkg/ctl"
+	"github.com/weaveworks/eksctl/pkg/eks"
+	"github.com/weaveworks/eksctl/pkg/eks/api"
+	"github.com/weaveworks/eksctl/pkg/utils"
+
+	awseks "github.com/aws/aws-sdk-go/service/eks"
+)
+
+func createNodeGroupCmd() *cobra.Command {
+	cfg := api.NewClusterConfig()
+	ng := cfg.NewNodeGroup()
+
+	cmd := &cobra.Command{
+		Use:   "nodegroup",
+		Short: "Create a nodegroup",
+		Run: func(_ *cobra.Command, args []string) {
+			if err := doAddNodeGroup(cfg, ng, ctl.GetNameArg(args)); err != nil {
+				logger.Critical("%s\n", err.Error())
+				os.Exit(1)
+			}
+		},
+	}
+
+	fs := cmd.Flags()
+
+	addCommonCreateFlags(fs, cfg, ng)
+	fs.StringVarP(&cfg.ClusterName, "name", "n", "", "Name of the EKS cluster to add the nodegroup to")
+
+	return cmd
+}
+
+func doAddNodeGroup(cfg *api.ClusterConfig, ng *api.NodeGroup, name string) error {
+	ctl := eks.New(cfg)
+
+	if cfg.Region != api.EKSRegionUSWest2 && cfg.Region != api.EKSRegionUSEast1 && cfg.Region != api.EKSRegionEUWest1 {
+		return fmt.Errorf("%s is not supported only %s, %s and %s are supported", cfg.Region, api.EKSRegionUSWest2, api.EKSRegionUSEast1, api.EKSRegionEUWest1)
+	}
+
+	if err := ctl.CheckAuth(); err != nil {
+		return err
+	}
+
+	if utils.ClusterName(cfg.ClusterName, name) == "" {
+		return fmt.Errorf("--name=%s and argument %s cannot be used at the same time", cfg.ClusterName, name)
+	}
+	cfg.ClusterName = utils.ClusterName(cfg.ClusterName, name)
+
+	if ng.SSHPublicKeyPath == "" {
+		return fmt.Errorf("--ssh-public-key must be non-empty string")
+	}
+
+	//TODO:  do we need to do the AZ stuff from create????
+
+	if err := ctl.EnsureAMI(ng); err != nil {
+		return err
+	}
+
+	if err := ctl.LoadSSHPublicKey(ng); err != nil {
+		return err
+	}
+
+	logger.Debug("cfg = %#v", cfg)
+
+	//TODO: is this check needed????
+	// Check the cluster exists and is active
+	eksCluster, err := ctl.DescribeControlPlane()
+	if err != nil {
+		return err
+	}
+	if *eksCluster.Status != awseks.ClusterStatusActive {
+		return fmt.Errorf("cluster %s status is %s, its needs to be active to add a nodegroup", *eksCluster.Name, *eksCluster.Status)
+	}
+	logger.Info("found cluster %s", eksCluster.Name)
+	logger.Debug("cluster = %#v", eksCluster)
+
+	{
+		stackManager := ctl.NewStackManager()
+		maxSeq, err := stackManager.GetMaxNodeGroupSeq()
+		if err != nil {
+			return err
+		}
+		ng.ID = maxSeq + 1
```

Comment:
This breaks after you create a nodegroup and then delete it. We'd better use a semi-random name other than an auto-incremented index. UUID can be used for example, but I'd use the same fancy names used for cluster names.

---

Let's say you had firstly created a nodegroup with ID=0(=maxSeq + 1), the subsequent `create nodegroup` sees maxSeq as 1! Which breaks assumptions in the codebase and results in index out of bound errors.

This happens because CloudFormation DescribeStacks API returns stacks in DELETE_COMPLETE for a while(maybe several to dozen minutes). See the `StackStatus: "DELETE_COMPLETE",` part of the example API response below:

```
2018-12-04T17:05:59+09:00 [▶]  stacks = [{
  Capabilities: ["CAPABILITY_IAM"],
  CreationTime: 2018-12-04 07:21:34.572 +0000 UTC,
  DeletionTime: 2018-12-04 07:49:49.204 +0000 UTC,
  Description: "EKS nodes (AMI family: AmazonLinux2, SSH access: false, subnet topology: Public) [created and managed by eksctl]",
  DisableRollback: false,
  Outputs: [{
      ExportName: "eksctl-adorable-sheepdog-1543570021-nodegroup-0::InstanceRoleARN",
      OutputKey: "InstanceRoleARN",
      OutputValue: "arn:aws:iam::ACCOUNT:role/eksctl-adorable-sheepdog-15435700-NodeInstanceRole-9D8J0JPOZOA"
    }],
  RollbackConfiguration: {

  },
  StackId: "arn:aws:cloudformation:us-west-2:ACCOUNT:stack/eksctl-adorable-sheepdog-1543570021-nodegroup-0/3aed7b30-f795-11e8-99dc-500c337110fd",
  StackName: "eksctl-adorable-sheepdog-1543570021-nodegroup-0",
  StackStatus: "DELETE_COMPLETE",
  Tags: [{
      Key: "eksctl.cluster.k8s.io/v1alpha1/cluster-name",
      Value: "adorable-sheepdog-1543570021"
    },{
      Key: "eksctl.cluster.k8s.io/v1alpha1/nodegroup-id",
      Value: "0"
    }]
}]
```

Reply:
Update: So I was able to fix the index out of bound errors just by replacing a bunch of `ng.ID` that are used to drill down `ClusterConfig` to a single node group inside it, to use `api.NodeGroup` instead.

I'll keep it going with `ng.ID = maxSeq + 1` for now.